packages/headless/src/controllers/triggers/headless-execute-trigger.ts

```diff
@@ -0,0 +1,90 @@
+import {TriggerSection} from '../../state/state-sections';
+import {triggers} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {logTriggerExecute} from '../../features/triggers/trigger-analytics-actions';
+import {SearchEngine} from '../../app/search-engine/search-engine';
+
+/**
+ * The `ExecuteTrigger` controller handles execute trigger actions.
+ */
+export interface ExecuteTrigger extends Controller {
+  /**
+   * the state of the `ExecuteTrigger` controller.
+   */
+  state: ExecuteTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `ExecuteTrigger` controller.
+ */
+export interface ExecuteTriggerState {
+  /**
+   * The name of the function to be executed.
+   */
+  name: string;
+
+  /**
+   * The parameters of the function to be executed.
+   */
+  params: {}[];
+}
+
+/**
+ * Creates a `ExecuteTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @returns A `RedirectionTrigger` controller instance.
+ * */
+export function buildExecuteTrigger(engine: SearchEngine): ExecuteTrigger {
+  if (!loadExecuteTriggerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const getState = () => engine.state;
+
+  let previousName = getState().triggers.execute.name;
+  let previousParams = getState().triggers.execute.params;
+
+  return {
+    ...controller,
+
+    subscribe(listener: () => void) {
+      const strictListener = () => {
+        const hasChanged =
+          previousName !== this.state.name ||
+          previousParams !== this.state.params;
+
+        previousName = this.state.name;
+        previousParams = this.state.params;
+
+        if (hasChanged && this.state.name) {
+          listener();
+          //const funct: keyof typeof window = this.state.name;
+          //window[funct](this.state.params);
+          //window[this.state.name as keyof window](this.state.params);
```

Comment:
I am trying to call the function with its parameters passed by a Trigger Execute object. The way I know how to call a JS function by name is by using the global window object.

However, I am getting some Typescript errors. I cannot use a string to index window such as `window[this.state.name]`. Also when trying to specifically declare it as a key, `keyof typeof window` does not accept strings.

I have found some posts online suggesting solutions: https://newbedev.com/typescript-ts7015-element-implicitly-has-an-any-type-because-index-expression-is-not-of-type-number. However I don't think any of these solutions are applicable here.

Reply:
We should not assume that the function is on the `window` object, because we want headless to be able to be used in any javascript runtime. In this case, typescript is correctly warning that the function might not be present on `window`.

I think this is okay, since a developer can use the controller like this:

```
executeTrigger.subscribe(() => {
    const {name, params} = executeTrigger.state;

    if (name=== 'add') {
         add(...params) // developer is responsible for referencing the add function, which may or may not be global
    }
})
```